include/pybind11/pytypes.h

```diff
@@ -1637,7 +1639,29 @@ class capsule : public object {
         }
     }
 
-    const char *name() const { return PyCapsule_GetName(m_ptr); }
+    const char *name() const { return get_name_or_throw(m_ptr); }
+
+    /// Replaces a capsule's name *without* calling the destructor on the existing one.
+    void set_name(const char *new_name) {
+        if (PyCapsule_SetName(m_ptr, new_name) != 0) {
+            throw error_already_set();
+        }
+    }
+
+private:
+    static const char *get_name_or_throw(PyObject *o) {
+        /* an exception may be in-flight, we must save it in case we create another one */
```

Comment:
Oh ... your code is great, but that we need it here isn't.

After thinking about it for a moment: it kind-of makes sense that it can happen in a capsule destructor (objects are deleted while exceptions unwind). But then we need to be super careful how we handle cascading errors, we don't want to mask the original error with the `PyCapsule_GetName()` error, that can be very confusing & waste a lot of time debugging (I lost a few days b/o masked errors only a couple months ago). I think to avoid that we need to chain the errors with `raise_from`.

Ideally the complex error handling should only be done from the destructors, not from the plain `name()` member function. It's a bug if `name()` is called while `PyErr_Occurred()` is true, unless it is from a destructor.

This is turning out far more complex than I would have thought. Maybe wait until @Skylion007 can comment on my suggestions before making more changes.

Reply:
Yes, I can wait, and your suggestions make perfect sense. 

I have pushed a change that initializes variables in `get_name_or_throw` but will wait to give @Skylion007 time to comment.

Reply:
I think you are absolutely correct, and `name()` should not call `get_name_or_throw`. 

Also the `get_name_or_throw` should use `raise_from` in the clause where call to `PyCapsule_GetName` raised an exception.

I will prepare the change, but will defer pushing it.

Reply:
Please go ahead with the push.
I just didn't want to rush you.
But certainly don't want to stop you from doing what you feel is a good direction.

Reply:
The [recommended practice](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_WriteUnraisable) for handling exceptions raised during execution of destructors is to use `PyErr_WriteUnraisable` passing the object being destroyed as it's argument.

This is simpler and more straightforward that trying to chain exceptions. The issue with chaining exceptions is that since I can only pass string error message to `raise_from`, we loose the traceback of the exception caused by `PyCapsule_Name`.
To turn `value` associated with this exception we need to call `PyObject_Repr` on it, which can raise additional exceptions. 

I am thus giving up on the chaining idea and will use `PyErr_WriteUnraisable` instead.

Reply:
Sounds awesome. I wasn't aware of `PyErr_WriteUnraisable` before. Great to know we're in a situation that's formally covered by the Python C API.

Reply:
I think the right thing to do if `PyCapsule_GetName` raises is to write unraisable, and then simply restore the error in flight (if any) and return the name. 

With this, `get_name_or_throw` should probably be `get_name_no_throw`.